src/program/reflection.rs

```diff
@@ -1017,3 +1017,164 @@ fn glenum_to_transform_feedback_mode(value: gl::types::GLenum) -> TransformFeedb
         v => panic!("Unknown value returned by OpenGL varying mode: {}", v)
     }
 }
+
+/// Contains all subroutine data of a program.
+#[derive(Debug, Clone)]
+pub struct SubroutineData {
+    /// Number of subroutine uniform locations per shader stage.
+    /// This is *not* equal to the number of subroutine uniforms per stage,
+    /// because users can use `#layout(location=...)`.
+    pub location_counts: HashMap<ShaderStage, usize>,
+
+    /// The list of all subroutine uniforms of the program stored in a structured way to enable fast lookups.
+    /// A subroutine uniform is uniquely defined by a name and a shader stage.
+    pub subroutine_uniforms: HashMap<(String, ShaderStage), SubroutineUniform>,
+}
+
+/// Information about a Subroutine Uniform (except name)
+#[derive(Debug, Clone)]
+pub struct SubroutineUniform {
+
+    /// The index of the subroutine uniform.
+    /// Needed to query information from the OpenGL backend.
+    pub index: u32,
+
+    /// The location of the uniform.
+    /// This is used to bind subroutines to this subroutine uniform.
+    pub location: i32,
+
+    /// A list of subroutines that can potentially be used with this uniform.
+    pub compatible_subroutines: Vec<Subroutine>,
+}
+
+/// Information about a subroutine.
+#[derive(Debug, Clone)]
+pub struct Subroutine {
+    /// The index of the subroutine, needed to bind this to a subroutine uniform.
+    pub index: u32,
+
+    /// The name of the subroutine.
+    pub name: String,
+}
+
+/// The different stages of the program pipeline.
+#[allow(missing_docs)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum ShaderStage {
+    Vertex,
+    Fragment,
+    TessellationControl,
+    TessellationEvaluation,
+    Geometry,
+    // FIXME? According to https://www.opengl.org/sdk/docs/man/html/glUniformSubroutines.xhtml ,
+    // compute shaders are not supported.
+    // Compute,
+}
+
+impl ShaderStage {
+    /// Converts the `ShaderStage` to its GLenum equivalent
+    pub fn to_gl_enum(&self) -> gl::types::GLenum {
+        match *self {
+            ShaderStage::Vertex => gl::VERTEX_SHADER,
+            ShaderStage::Fragment => gl::FRAGMENT_SHADER,
+            ShaderStage::TessellationControl => gl::TESS_CONTROL_SHADER,
+            ShaderStage::TessellationEvaluation => gl::TESS_EVALUATION_SHADER,
+            ShaderStage::Geometry => gl::GEOMETRY_SHADER,
+            // Compute => gl::COMPUTE_SHADER,
+        }
+    }
+}
+
+fn get_supported_shader_stages(ctxt: &mut CommandContext) -> Vec<ShaderStage> {
+    let mut stages = vec![ShaderStage::Vertex, ShaderStage::Fragment];
+    if ctxt.extensions.gl_arb_tessellation_shader {
+        stages.push(ShaderStage::TessellationControl);
+        stages.push(ShaderStage::TessellationEvaluation);
+    }
+    if ctxt.extensions.gl_arb_geometry_shader4 {
```

Comment:
There's a third awkward solution, which is that glium could add a small header when returning a program binary and parse it when reloading.

OpenGL implementations are free to refuse to load any binary for whatever reason they want, so technically it would be a valid thing to do.